### ISSUE TITLE
data fetchers call static methods

### ISSUE THREAD

**Issue by softprops** [graphql-java/graphql-java#3247]
**Describe the bug**

I've recently attempted to upgrade to v20 and noticed an unexpected behavior

Thankfully caught by our unit tests 

a snipped from our unit tests output showed that graphql java failed to serialize a proper and valid response

> updateOrganizerSubscriptionSuccessPromoCode() ✘ expected: <{data={updateOrganizerSubscription={success=true, error=null}}}> but was: <{data={updateOrganizerSubscription={success=null, error=null}}, errors=[{message=Can't serialize value (/updateOrganizerSubscription/success) : Expected a value that can be converted to type 'Boolean' but it was a 'UpdateOrganizerSubscriptionPayload', path=[updateOrganizerSubscription, success], extensions={classification=DataFetchingException}}]}>

the schema in question looks like this

```graphql
type UpdateOrganizerSubscriptionPayload {
  """
  A payload in case an organizer subscription was successfully updated.
  """
  success: Boolean

  """
  A payload in case updating an organizer subscription failed.
  """
  error: OrganizerSubscriptionError
}
```

the data fetcher returned class of the that confirms to this 

```java
public class UpdateOrganizerSubscriptionPayload {
  private final Boolean success;
  private final OrganizerSubscriptionError error;

  private UpdateOrganizerSubscriptionPayload(Boolean success, OrganizerSubscriptionError error) {
    this.success = success;
    this.error = error;
  }

  public static UpdateOrganizerSubscriptionPayload success() { // 👈 note the static factory method for creating a success payload
    return new UpdateOrganizerSubscriptionPayload(Boolean.TRUE, null);
  }

  public static UpdateOrganizerSubscriptionPayload error(OrganizerSubscriptionError error) {  // 👈 note the static factory method for creating a success payload
    return new UpdateOrganizerSubscriptionPayload(null, error);
  }

  public Boolean getSuccess() {
    return success;
  }

  public OrganizerSubscriptionError getError() {
    return error;
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) return true;
    if (o == null || getClass() != o.getClass()) return false;
    UpdateOrganizerSubscriptionPayload that = (UpdateOrganizerSubscriptionPayload) o;
    return Objects.equals(getSuccess(), that.getSuccess())
        && Objects.equals(getError(), that.getError());
  }

  @Override
  public int hashCode() {
    return Objects.hash(getSuccess(), getError());
  }

  @Override
  public String toString() {
    return new StringJoiner(
            ", ", UpdateOrganizerSubscriptionPayload.class.getSimpleName() + "[", "]")
        .add("success=" + success)
        .add("error=" + error)
        .toString();
  }
}
```

but the serialization error I get is 

> Expected a value that can be converted to type 'Boolean' but it was a 'UpdateOrganizerSubscriptionPayload'

I believe this relates to the Record Like Property Fetching Support section of https://github.com/graphql-java/graphql-java/releases/tag/v20.0

I believe what is happening is that the new record like property datafetcher is reflecting over the name `success` but this is not expected as its static method intended to build a "success" version of the data fetcher payload. previously the `getSuccess` method was resolved. The release notes call out that the record like property name is preferred so the factory method likely gets picked up and instead of a Boolean indeed I get the type which contains the boolean.

I'm not sure if this is a bug or a feature but I feel like the issue and resolution might be to exclude methods with a `static` modifier when lifting the property value for serializing the result.

**To Reproduce**
Please provide a code example or even better a test to reproduce the bug.

I don't have a more limited case but I suspect if you did the following you could reproduce the issue

schema 

```graphql
type Test {
  success: Boolean
}
```

data fetcher 

```java
class TestFetcher implements DataFetcher<TestFetcher.Test> {
   public static class Test {
       private boolean success;
       public Test(boolean success) {
         this.success = success;
      }
     public boolean getSuccess() { return this.success }  // 👈  expect this
     public static Test success() { return new Test(true); }  // 👈  likely will use this by fail when validating schema
  }
  public Test get(DataFetchingEnvironment env) {
     return Test.success();
  }
}
```

again I didn't look but likely the fix for expected behavior in my case with be to omit methods with `static` modifiers when selecting candidate methods to invoke

**Comment by softprops** [comment]
We were able to work around this by removing all potentially risky static factory methods in datafetcher response types but I’m leaving the issue open in case it might help remove some upgrade surprises for others

**Comment by bbakerman** [comment]
Thanks for the detailed report.  We will look into this and try to make a reproduction test.

We should NOT be reading properties from static methods and I will check that this is still the case.

I have some problems with your reproduction classes and error messages however.

> Expected a value that can be converted to type 'Boolean' but it was a 'UpdateOrganizerSubscriptionPayload

This error message says that the object passed into the Boolean scalar was one of class `UpdateOrganizerSubscriptionPayload` - I suspect you have some level problems here with your code..  eg you think that are returning a `success` boolean but somehow returning the full object??

And this repro `TestFetcher` is funky.  That is if its a data fetcher class for a field then `get()` will be called and whatever is returned is used.  in your case you called a `.success()` method via code.   So thats what will be used.

```
 public Test get(DataFetchingEnvironment env) {
     return Test.success();
  }
```

I think you example would not compile actually - its said to return `Test` but your `success()` method is a `boolean`.  You maybe have been typing an example from memory rather than from a compiler however.

I will say that for any given field - if its backed by a DataFetcher specifically - then whatever it returns will be used.

If there is no mapping them a default `PropertyDataFetcher` will be used and that should use record / POJO accessor methods and I will check that they still do or not

**Comment by bbakerman** [comment]
NOTE : I changed the name of the issue - yes it was a regression for you - in that record like static methods now get called - so its a regression from your point of view.

But the idea that it calls static methods are all is troubling to me and unexpected.  So I want to describe the problem in broader terms

**Comment by bbakerman** [comment]
What I plan to do about this issue is this

Right now its impossible to have both a static and non static method in a Java class

```
class X {
   public String getFoo() {...}
   public static String getFoo() {...}
}
```

So if someone was getting the static in the old days - then  they cant be expecting the non static method as well since it does not compile.

Today the search for candidate methods are
* search public `recordLike()` methods
   * could be static or non static
* search public `getFoo()` methods
   * could be static or non static
   
So I propose we do this during method candidate search in a future version

* search public non static `recordLike()` methods
* search public non static `getFoo()` methods
* search public static `recordLike()` methods
* search public static `getFoo()` methods

So we go for instance methods first but if you had the old static behavior - you will still get it because it will find your static AND it cant get confused - since you cant have both today in Java 

This would fix the regression above since it would prefer instance methods over static methods first

**Comment by dondonz** [comment]
Closed by #3251 (and included in today's v21.0 release)
